src/db.rs

```diff
@@ -64,36 +65,39 @@ pub struct SingleThreaded {
 ///
 /// See [`DB`] for more details, including performance implications for each mode
 pub struct MultiThreaded {
-    cfs: RwLock<BTreeMap<String, ColumnFamily>>,
+    cfs: RwLock<BTreeMap<String, Arc<UnboundColumnFamily>>>,
 }
 
 impl ThreadMode for SingleThreaded {
-    fn new(cfs: BTreeMap<String, ColumnFamily>) -> Self {
-        Self { cfs }
+    fn new(cfs: BTreeMap<String, *mut ffi::rocksdb_column_family_handle_t>) -> Self {
+        Self {
+            cfs: cfs
+                .into_iter()
+                .map(|(n, c)| (n, ColumnFamily { inner: c }))
+                .collect(),
+        }
     }
 
     fn cf_drop_all(&mut self) {
-        for cf in self.cfs.values() {
-            unsafe {
-                ffi::rocksdb_column_family_handle_destroy(cf.inner);
-            }
-        }
+        // Cause all ColumnFamily objects to be Drop::drop()-ed.
+        self.cfs.clear();
     }
 }
 
 impl ThreadMode for MultiThreaded {
-    fn new(cfs: BTreeMap<String, ColumnFamily>) -> Self {
+    fn new(cfs: BTreeMap<String, *mut ffi::rocksdb_column_family_handle_t>) -> Self {
```

Comment:
Can't we create `ffi::rocksdb_column_family_handle_t` inside the method `new` and then a signature of the method could be changed to `fn new(cfs: BTreeSet<String>) -> Self` ? 🤔

Reply:
technically, yes. but I double its merit...

As I went long explaining here  (https://github.com/rust-rocksdb/rust-rocksdb/commit/0d5ff023b42536aee8869fac5a504b553ff0fe4d), this trait method is called deep inside DB construction at (`open_cf_descriptors_internal`). So, we have to move large part of code inside `ThreadMode`.. Then `ThreadMode` will be significantly fatter than now, although it's not intended to be used for users still as was before. Then this only gives us the ability of getting rid of those `ffi::` types in strictly _public_ APIs.